Redirect shard-level bulk failures to a failure store if applicable

jbaiera opened the pull request:
When indexing data from something like a fleet integration, a mapping change that is incompatible with the incoming data often results in the mass rejection of documents with little to no feedback to the end user. While a data shipper may be receiving error responses, there is little that the shipper can do about them, and thus the data is often dropped. The issue is usually only discovered by an end user after the realization that little to no data has been ingested since the mapping changes, and often still requires more effort still to diagnose the actual mapping problem.

This PR expands upon previous work in the failure store by inspecting failed shard-level bulk operations and possibly redirecting them to a failure store. Documents are eligible to have their failures stored if they are writing to a data stream with a failure store enabled. Responses from the TransportShardBulkAction are saved in the working bulk response list, but any failures that are eligible for redirection will be transformed into failure documents and retried. The second attempt for storing faliures will overwrite the original failure on the working response.

jbaiera committed:
Add javadoc for ConcreteIndices

jbaiera committed:
Extract onBulkComplete runnable into its own method.

jbaiera committed:
Add iterator implementation that zips items with their index.

jbaiera committed:
Add ability to adapt a supplier into an iterator.

This is useful to allow us to iterate and remove the contents of a queue in a uniform way that mirrors iterating over a list without removals.

jbaiera committed:
Clean up generics warning.

Type argument T is redundant.

jbaiera committed:
Modify BulkOperation to redirect shard level failures to a failure store if applicable.

jbaiera committed:
Update docs/changelog/105362.yaml

elasticsearchmachine commented:
Hi @jbaiera, I've created a changelog YAML for you.

jbaiera commented on server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:
```diff
@@ -282,6 +480,8 @@ public void onNewClusterState(ClusterState state) {
 
             @Override
             public void onClusterServiceClose() {
+                // TODO: There is likely very little we can do about this, and our time in this JVM is likely short,
+                //  should we just get out of here ASAP?
```
As mentioned in the comment: There is some weirdness about this edge case that gets introduced when we have two phases of checks.

Originally, if we reached this code, it was before we submitted any operations to the shards, and thus simply failing the whole request is pretty reasonable since it has not even really started yet.

Now, we have a chance of encountering this after we've written our data to the shards, but there were failures that we want to redirect. Simply failing the entire request here feels wrong, since it has mostly succeeded, the only problem is that the failure store feature has not kicked in yet. If the bulk operation had updates mixed into it, the client might see those updates as having not been applied when they have indeed been applied before the cluster service closed.

I'm looking into what is going on when this kind of case happens. My suspicion is that the node is being shut down and we're not long for the world. It would be simple enough to have this call the new `onClusterBlocked` handler, which will be identical logic in the first case above, but in the new case it will simply skip trying to send any failure store redirects and return the results of the bulk (failures included) immediately.

jbaiera committed:
Fix changelog category

jbaiera committed:
Merge branch 'main' into data-stream-shard-failure-redirect

jbaiera committed:
Add tests for BulkOperation

jbaiera committed:
precommit tomfoolery

jbaiera committed:
Add rest test

jbaiera committed:
Add iterators tests

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

elasticsearchmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

dakrone commented on server/src/test/java/org/elasticsearch/action/bulk/BulkOperationTests.java:
```diff
@@ -0,0 +1,793 @@
+    }
+
+    /**
+     * If a bulk operation times out while waiting for cluster blocks to be cleared, it should fail the request.
```
Thanks for commenting all of these!

dakrone commented on server/src/test/java/org/elasticsearch/action/bulk/BulkOperationTests.java:
```diff
@@ -0,0 +1,793 @@
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+public class BulkOperationTests extends ESTestCase {
```
Would it be possible to add a test that fails to transform the failure store doc (throws an `IOException` in `addDocumentToRedirectRequests`) to make sure that we do the right thing?

dakrone reviewed (changes requested):
Thanks for working on this Jimmy! This looks really close to me, I left only a few comments

jbaiera commented on server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:
```diff
@@ -282,6 +531,8 @@ public void onNewClusterState(ClusterState state) {
 
             @Override
             public void onClusterServiceClose() {
+                // TODO: There is likely very little we can do about this, and our time in this JVM is likely short,
+                // should we just get out of here ASAP?
```
So, this affects regular indexing, but if it reaches this point for regular indexing the entire operation fails and we move on. If this happens during the failure store indexing section, then we may still have accepted some of the documents in the request. It's a question of whether or not it is the right move to simply reject the failure store parts and return the request with the failures or if it is better to just fail the entire request since we may be racing with the JVM shutting down.

Or at least I assume we're racing with a JVM that is shutting down, I haven't explored the intricacies of why else the cluster service might be closed.

gmarouli replied:
Even if there are other reasons for the cluster service shutting down, it's safe to assume that multiple things will not be working anymore and they are not coming back. So, I think to start with we could fail the request.

jbaiera replied:
I have removed the TODO marker but left most of the comment intact. In the case that this is the regular indexing, we continue with the logic as it existed before: We immediately fail the request before doing any work. In the event that the cluster closes while we are about to send the failure information, then we once again will simply fail the entire request. The second phase will not be successful anyway, and we should free up the threads without any delay to allow for a timely shutdown.

jbaiera commented on server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:
```diff
@@ -97,68 +138,88 @@ final class BulkOperation extends ActionRunnable<BulkResponse> {
+    }
+
+    private void doRedirectFailures() {
+        assert failureStoreRedirects.isEmpty() != true : "Attempting to redirect failures, but none were present in the queue";
```
I think that makes sense, though I think it might be better to check that during the failure handling steps as well so that we aren't queuing data that won't be needed.

jbaiera replied:
Aha, we do check it at that point already in `org.elasticsearch.action.bulk.BulkOperation#getRedirectTarget`

I'll add one more check before doing the waiting for sanity's sake though.

jbaiera replied:
I added the check to `redirectFailuresOrCompleteBulkOperation` which calls into this method.

jbaiera committed:
Addressing review comments

nielsbauman commented on server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:
```diff
@@ -97,68 +138,88 @@ final class BulkOperation extends ActionRunnable<BulkResponse> {
+    private void doRedirectFailures() {
+        assert failureStoreRedirects.isEmpty() != true : "Attempting to redirect failures, but none were present in the queue";
+        final ClusterState clusterState = observer.setAndGetObservedState();
+        if (handleBlockExceptions(clusterState, ActionRunnable.run(listener, this::doRedirectFailures), this::discardRedirectsAndFinish)) {
```
I assume we discard redirects because we don't want to start doing more stuff (i.e. the failure store writes) when the cluster is blocked? Is that maybe worth a comment or do you think that's clear enough already for non-rookies?

jbaiera replied:
I have added a javadoc on `handleBlockException` and a comment on this line to make it more clear

nielsbauman commented on server/src/main/java/org/elasticsearch/common/collect/Iterators.java:
```diff
@@ -225,6 +227,66 @@ public U next() {
         }
     }
 
+    public static <T, U> Iterator<U> zip(Iterator<? extends T> input, BiFunction<Integer, T, ? extends U> fn) {
```
Any particular reason you chose to call this `zip`? I know `zip` as a function that "zips" _two or more_ iterators. This looks more like a `map` iterator to me (but with the index passed). Would it maybe make more sense to rename this to `mapWIthIndex` or something along those lines?

jbaiera replied:
These are great points, my memory is flawed. Scala calls this `zipWithIndex`, which is what I was basically wanting.

I went back and forth between `mapWithIndex`, `zipWithIndex` and `enumerate` but I ended up going with `enumerate` in the end since we already have a `map` call, and may at some point want to add a proper `zip` iterator.

nielsbauman commented on server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:
```diff
@@ -239,33 +385,136 @@ public void onFailure(Exception e) {
+                        addDocumentToRedirectRequests(request, e, failureStoreReference);
+                    }
+
+                    // Check for any previously recorded responses in the buffer since this might be a failure during a redirect
+                    BulkItemResponse response = responses.get(request.id());
+                    if (response == null) {
+                        BulkItemResponse.Failure failure = new BulkItemResponse.Failure(indexName, docWriteRequest.id(), e);
+                        response = BulkItemResponse.failure(request.id(), docWriteRequest.opType(), failure);
+                    } else {
+                        // Response already recorded. We should only be here if the existing response is a failure and
+                        // we are encountering a new failure while redirecting.
+                        assert response.isFailed() : "Attempting to overwrite successful bulk item result with a failure";
+                        response.getFailure().getCause().addSuppressed(e);
+                    }
+                    // Always replace the item in the responses for thread visibility of any mutations
+                    responses.set(request.id(), response);
```
Is there a reason you don't use `BulkOperation#addFailure` here? It seems to match that method, except for the last line of the method.

jbaiera committed:
Add a test for when failure store documents cannot be converted.

Refactor converter code to be injectable to the operation for testing.
Some additional test cleanup.

jbaiera committed:
Change "zip" iterators call to "enumerate"

jbaiera commented on docs/changelog/105362.yaml:
```diff
@@ -0,0 +1,5 @@
+pr: 105362
+summary: Redirect shard-level bulk failures to a failure store if applicable
```
Hmm, this is news to me, I can mark it as such but the previous PR's I've put forward are also not `>non-issue`

gmarouli replied:
Nvm then :)

jbaiera committed:
Some late refactors based on review comments.

Added more javadocs in some confusing places.

Combined failure handling logic down to one or two methods instead of repeating it in multiple places.

Downgrade the TODO on the cluster service close call.

jbaiera committed:
Precommit

jbaiera committed:
Fixing test bug after class rename

jbaiera committed:
Merge branch 'main' into data-stream-shard-failure-redirect

jbaiera committed:
Delete docs/changelog/105362.yaml

gmarouli commented on server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:
```diff
@@ -239,33 +381,135 @@ public void onFailure(Exception e) {
+            // The index abstraction could be an alias. Alias abstractions (even for data streams) only keep track of which _index_ they
+            // will write to, not which _data stream_.
+            // We work backward to find the data stream from the concrete write index to cover this case.
+            Index concreteIndex = ia.getWriteIndex();
+            IndexAbstraction writeIndexAbstraction = metadata.getIndicesLookup().get(concreteIndex.getName());
+            DataStream parentDataStream = writeIndexAbstraction.getParentDataStream();
+            if (parentDataStream != null && parentDataStream.isFailureStore()) {
+                // Keep the data stream name around to resolve the redirect to failure store if the shard level request fails.
+                return parentDataStream.getName();
+            }
```
Food for thought, I think we might need this more often as a follow up PR. Maybe should add this in the `IndexAbstraction`, I have a feeling it's not the first time we do something like that.

gmarouli reviewed (commented):
Hey @jbaiera, I didn't have time to go through the whole PR. It does look really good though. Hopefully tomorrow I will finish it.

jbaiera committed:
Typo

jbaiera commented on modules/data-streams/src/yamlRestTest/resources/rest-api-spec/test/data_stream/190_failure_store_redirection.yml:
```diff
@@ -108,3 +108,83 @@ teardown:
+  - skip:
+      version: " - 8.13.99"
+      reason: "data stream failure stores only redirect ingest failures in 8.14+"
+      features: [allowed_warnings, contains]
```
This comment needed an update! Should accurately be about shards instead of ingest now

gmarouli reviewed (approved):
I am going to give a slightly reserved LGTM, considering the area of expertise that's as good as it gets 🚀 .

This is a big improvement on existing code as well. Thank you @jbaiera !

jbaiera commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into data-stream-shard-failure-redirect
